ocaml/xapi-storage-script/main.ml

```diff
@@ -372,168 +678,152 @@ let fork_exec_rpc :
     -> ?compat_in:compat_in
     -> ?compat_out:compat_out
     -> R.call
-    -> R.response Deferred.t =
+    -> R.response Lwt.t =
  fun ~script_dir ?missing ?(compat_in = id) ?(compat_out = id) ->
   let invoke_script call script_name :
-      (R.response, Storage_interface.Errors.error) Deferred.Result.t =
-    Process.create ~prog:script_name ~args:["--json"] () >>= function
-    | Error e ->
-        error "%s failed: %s" script_name (Error.to_string_hum e) ;
-        return
-          (Error
-             (backend_error "SCRIPT_FAILED" [script_name; Error.to_string_hum e])
+      (R.response, Storage_interface.Errors.error) Lwt_result.t =
+    (* We pass just the args, not the complete JSON-RPC call.
+       Currently the Python code generated by rpclib requires all params to
+       be named - they will be converted into a name->value Python dict.
+       Rpclib currently puts all named params into a dict, so we expect
+       params to be a single Dict, if all the params are named. *)
+    ( match call.R.params with
+    | [(R.Dict _ as d)] ->
+        return d
+    | _ ->
+        fail
+          (backend_error "INCORRECT_PARAMETERS"
+             [
+               script_name
+             ; "All the call parameters should be named and should be in a RPC \
+                Dict"
+             ]
           )
-    | Ok p -> (
+    )
+    >>>= fun args ->
+    let args = compat_in args in
+    let p = Process.create ~prog:script_name ~args:["--json"] in
+    Lwt.catch
+      (fun () ->
         (* Send the request as json on stdin *)
-        let w = Process.stdin p in
-        (* We pass just the args, not the complete JSON-RPC call.
-           Currently the Python code generated by rpclib requires all params to
-           be named - they will be converted into a name->value Python dict.
-           Rpclib currently puts all named params into a dict, so we expect
-           params to be a single Dict, if all the params are named. *)
-        ( match call.R.params with
-        | [(R.Dict _ as d)] ->
-            return (Ok d)
-        | _ ->
-            return
-              (Error
-                 (backend_error "INCORRECT_PARAMETERS"
-                    [
-                      script_name
-                    ; "All the call parameters should be named and should be \
-                       in a RPC Dict"
-                    ]
-                 )
-              )
+        let w = p.proc#stdin in
+        Lwt_io.write w (Jsonrpc.to_string args) >>= fun () -> Lwt_io.close w
+      )
+      (fun _ -> Process.close p >>= fun _ -> Lwt.return_unit)
+    >>= fun () ->
+    Process.collect_output_and_wait p >>= fun output ->
```

Comment:
Can we use `Lwt_process.pmap` to simplify some of this?

Reply:
I'll try, but that method doesn't handle stderr

Reply:
I don't believe it's usable, there's no way to extract the exit status when using pmap